icu4j/main/classes/core/src/com/ibm/icu/impl/units/UnitConverter.java

```diff
@@ -0,0 +1,309 @@
+// © 2020 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+
+package com.ibm.icu.impl.units;
+
+import com.ibm.icu.util.MeasureUnit;
+
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.regex.Pattern;
+
+import static java.math.MathContext.DECIMAL128;
+
+public class UnitConverter {
+    private BigDecimal conversionRate;
+    private BigDecimal offset;
+
+    /**
+     * Constructor of `UnitConverter`.
+     * NOTE:
+     * - source and target must be under the same category
+     * - e.g. meter to mile --> both of them are length units.
+     *
+     * @param source          represents the source unit.
+     * @param target          represents the target unit.
+     * @param conversionRates contains all the needed conversion rates.
+     */
+    public UnitConverter(MeasureUnitImpl source, MeasureUnitImpl target, ConversionRates conversionRates) {
+        Convertibility convertibility = extractConvertibility(source, target, conversionRates);
+        assert (convertibility == Convertibility.CONVERTIBLE || convertibility == Convertibility.RECIPROCAL);
+
+        Factor sourceToBase = conversionRates.getFactorToBase(source);
+        Factor targetToBase = conversionRates.getFactorToBase(target);
+
+        if (convertibility == Convertibility.CONVERTIBLE) {
+            this.conversionRate = sourceToBase.divide(targetToBase).getConversionRate();
+        } else {
+            this.conversionRate = sourceToBase.multiply(targetToBase).getConversionRate();
+        }
+
+        // calculate the offset
+        this.offset = conversionRates.getOffset(source, target, sourceToBase, targetToBase, convertibility);
+    }
+
+    static public Convertibility extractConvertibility(MeasureUnitImpl source, MeasureUnitImpl target, ConversionRates conversionRates) {
+        ArrayList<SingleUnitImpl> sourceSingleUnits = conversionRates.getBasicUnitsWithoutSIPrefix(source);
+        ArrayList<SingleUnitImpl> targetSingleUnits = conversionRates.getBasicUnitsWithoutSIPrefix(target);
+
+        HashMap<String, Integer> dimensionMap = new HashMap<>();
+
+        insertInMap(dimensionMap, sourceSingleUnits, 1);
+        insertInMap(dimensionMap, targetSingleUnits, -1);
+
+        if (areDimensionsZeroes(dimensionMap)) return Convertibility.CONVERTIBLE;
+
+        insertInMap(dimensionMap, targetSingleUnits, 2);
+        if (areDimensionsZeroes(dimensionMap)) return Convertibility.RECIPROCAL;
+
+        return Convertibility.INCONVERTIBLE;
```

Comment:
Was using "inconvertible" instead of "unconvertible" a conscious choice? Both inconvertible (here) and unconvertible (C++) are valid English.

"1. unconvertible - used especially of currencies; incapable of being exchanged for or replaced by another currency of equal value. inconvertible, unexchangeable."

Reply:
I had a pull request before to change `UNCONVERTIBLE` in C++ to `INCONVERTIBLE` , but I do not see it.

Yes, it was a conscious choice. :)

Reply:
but flipped to `UNCONVERTIBLE`

Reply:
Thanks!